crates/core/src/protocols/evalcheck/subclaims.rs

```diff
@@ -328,33 +332,28 @@ where
 	projected_bivariate_claims
 		.par_iter()
 		.zip(metas)
-		.map(|(claim, meta)| {
-			match meta.inner_id {
-				Some(inner_id) => {
-					{
-						// packed / shifted
-						let inner_multilin = witness_index.get_multilin_poly(inner_id)?;
-						let eval_point = &claim.eval_point[meta.projected_n_vars..];
-						let query = memoized_queries
-							.full_query_readonly(eval_point)
-							.ok_or(Error::MissingQuery)?;
-						Ok(Some(
-							backend
-								.evaluate_partial_high(&inner_multilin, query.to_ref())
-								.map_err(Error::from)?,
-						))
-					}
-				}
-				None => Ok(None), // composite
+		.map(|(claim, meta)| match (meta.inner_id, meta.projected_id) {
+			(Some(inner_id), Some(_)) => {
+				let inner_multilin = witness_index.get_multilin_poly(inner_id)?;
+				let eval_point = &claim.eval_point[meta.projected_n_vars..];
+				let query = memoized_queries
+					.full_query_readonly(eval_point)
+					.ok_or(Error::MissingQuery)?;
+				Ok(Some(
+					backend
+						.evaluate_partial_high(&inner_multilin, query.to_ref())
```

Comment:
This partial evaluation doesn't seem to be added to the memoized map? Am I missing something?

Reply:
^ This is major. Reusing this computation is the main purpose of the issue.

Reply:
yes, but at the end of evalcheck we build map `(id, eval_point) -> MultilinearWitness`, and for ring_switch it's just result of `evaluate_partial_high`, just to avoid duplication of evals